Assets/MixedRealityToolkit.SDK/Experimental/Features/Dwell/DwellHandler.cs

```diff
@@ -0,0 +1,210 @@
+
+        /// <summary>
+        /// Cached pointer reference to track focus events maps to teh same pointer that initiated dwell
+        /// </summary>
+        private IMixedRealityPointer pointerRef;
+
+        private DateTime focusEnterTime = DateTime.MaxValue;
+        private DateTime focusExitTime = DateTime.MaxValue;
+
+        protected DwellState currentDwellState = DwellState.None;
+        protected float fillTimer = 0;
+
+        private void Awake()
+        {
+            Debug.Assert(dwellProfile != null, "DwellProfile is null, creating default profile.");
+
+            if (dwellProfile == null)
+            {
+                dwellProfile = ScriptableObject.CreateInstance("DwellProfile") as DwellProfile;
+            }
+
+            CancelDwellAction += OnDwellCancelAction;
+        }
+
+        private void Update()
+        {
+            if (HasFocus && currentDwellState != DwellState.DwellCompleted)
+            {
+                double focusTime = (DateTime.UtcNow - this.focusEnterTime).TotalSeconds;
+
+                if (focusTime >= dwellProfile.DwellIntentDelay && currentDwellState == DwellState.FocusGained)
+                {
+                    DwellIntended.Invoke(pointerRef);
+                    currentDwellState = DwellState.DwellIntended;
+                    fillTimer = 0;
+                }
+                else if ((focusTime - dwellProfile.DwellIntentDelay) >= dwellProfile.DwellStartDelay && currentDwellState == DwellState.DwellIntended)
+                {
+                    DwellStarted.Invoke(pointerRef);
+                    currentDwellState = DwellState.DwellStarted;
+                    fillTimer = dwellProfile.TimeToCompleteDwell;
+                }
+                else if ((focusTime - dwellProfile.DwellIntentDelay - dwellProfile.DwellStartDelay) >= dwellProfile.TimeToCompleteDwell && currentDwellState == DwellState.DwellStarted)
+                {
+                    Debug.Assert(fillTimer < float.Epsilon, $"Fill timer was not 0. It was {fillTimer}.");
+                    DwellCompleted.Invoke(pointerRef);
+                    currentDwellState = DwellState.DwellCompleted;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Default dwell progress calculation logic. Can be overriden by developers for customized visuals
+        /// </summary>
+        /// <returns>dwell progress as a float between 0 and 1</returns>
+        public virtual float CalculateDwellProgress()
+        {
+            float dwellProgress = 0;
+
+            switch (currentDwellState)
+            {
+                case DwellState.None:
+                    dwellProgress = 0;
+                    break;
+                case DwellState.FocusGained:
+                    dwellProgress = 0;
+                    fillTimer = 0;
+                    break;
+                case DwellState.DwellStarted:
+                    fillTimer -= Time.deltaTime;
```

Comment:
>Time.deltaTime [](start = 33, length = 14)

Does this imply that CalculateDwellProgress must be called exactly once per frame?  If so, then maybe do the math in Update() instead?

Reply:
> > Time.deltaTime [](start = 33, length = 14)
> 
> Does this imply that CalculateDwellProgress must be called exactly once per frame? If so, then maybe do the math in Update() instead?

Not necessarily. It is up to the caller to determine how often to make this call.